### PR TITLE
Rudimentary contact sharing

### LINKED ISSUE #68: Store personal profile in local db
- Maybe in a different box or in the same contacts box for now
- Will help with all the logic for displaying and sharing profile
- Different box would help store "saved templates" too
- Related to #25

### LINKED ISSUE #154: Create cloud function for sharing
blocked on #22

### LINKED ISSUE #163: Screen to display a received shared contact
with options to save and follow if enabled

### PR DESCRIPTION
**Pull request by TortoiseWrath**
Closes #163 

Closes #68 

Closes #154 

Is an improvement toward #165 and #25 

Does not really help with #164

### CODE REVIEW

**Comment by mayank99** [comment]
> Does it work?

**Comment by TortoiseWrath** [comment]
it works on my machine

**Comment by mayank99** [comment]
Can you link the issues it closes?

**Comment by TortoiseWrath** [comment]
oh yeah for sure

**Comment by mayank99** [comment]
Also how relevant is this commit message still?

> This is far from complete:
> * Name is saved in profile but does not appear on profile page
>   when loaded
> * App crashes when attempting to display share screen with QR code & URL
> * Need a share screen that is less than 100% bad
> * Numerous other shortcomings and shortcuts - see TODOs
> 
> And the database schema is bare minimum right now.
> 
> But at this point by pressing the share button (not the decoy share
> button) the profile is actually saved to firestore so we can start
> working on displaying it (#163, #164)
> 
> Changes in this commit:
> * Save contacts to database after adding/editing
>   * Newly added contacts gain a randomly-generated id and are not saved
>     to the base contacts app
>   * TODO: Find a way to save them to the base contacts app and
>     keep synchronized
> * Add isProfile flag to add/edit page and use as profile screen (#25)
>   * TODO: Needs button functionality moved to top-level button somehow
>   * Add shareProfile method
>     * Calls new static ContactSender.send
> 	  * Calls firebase function (#154, see previous commits)
> 	  * Returns shareable url
> 	* Displays new SharePage
> 	  * TODO: This does not work yet
> 	  * This share page will eventually be the FINAL page in a multi-
>         page sharing workflow (#165)
> * Save profile to contacts box with key 'profile' (#68)
> * Handle null values when converting MultaccContact to map
> * Get profile contact on load
> * Remove "profile" contact from main contacts screen

**Comment by TortoiseWrath** [comment]
that commit message deals entirely with sharing stuff, which this is branched off of. Most of that is still true apart from the UI stuff you fixed. I have a couple minor improvements to sharing coming

File: lib/pages/contacts/contacts_page.dart [hunk 1]
```diff
@@ -29,7 +29,9 @@ class _ContactsPageState extends State<ContactsPage> with WidgetsBindingObserver
   Widget build(BuildContext context) {
     return Observer(
       builder: (_) {
-        contactsData.allContacts.sort((a, b) => a.name.compareTo(b.name));
+        contactsData.allContacts
+          ..removeWhere((contact) => contact.clientKey == 'profile')
```
Comment (mayank99) [inline lib/pages/contacts/contacts_page.dart:33]: Don't need this anymore. See #210

**Comment by TortoiseWrath** [comment]
This works well enough to demo except you can't copy the link from the sharing page so you will have to switch to another app with a link in it and click it there (i.e. demonstrate sending and receiving separately)

File: lib/database/contact_model.dart [hunk 2]
```diff
@@ -95,12 +95,35 @@ class MultaccContact extends Contact {
         listEquals(this.emails?.toList(), baseContact.emails?.toList()) &&
         this.familyName == baseContact.familyName &&
         this.givenName == baseContact.givenName &&
-        this.identifier == baseContact.identifier &&
+        // this.identifier == baseContact.identifier &&
```
Comment (mayank99) [inline lib/database/contact_model.dart:98]: why is this commented out again?

Comment (TortoiseWrath) [reply lib/database/contact_model.dart:98]: uhhhhhh I don't remember but it works now

Comment (TortoiseWrath) [reply lib/database/contact_model.dart:98]: open an issue to figure out why this is

**Comment by mayank99** [comment]
Can you force push your last commit on top of this? #219 will take care of the leftover bs

**Comment by mayank99** [comment]
Squash time

**Comment by TortoiseWrath** [comment]
no don't

**Pull request merged**